Proposed WIT for tracing portion of wasi-observe

calebschoepp opened the pull request:
This PR proposes WIT for what I think the tracing portion of wasi-observe should look like. Initial discussion of high level approaches was had [here](https://github.com/WebAssembly/wasi-observe/pull/13) and this WIT is the outcome of that. I have about 50% of a working host implementation for this WIT in Spin so I'm pretty confident this is a workable approach implementation-wise.

Here are some of the design principles I followed while writing this WIT:
- Closely follow the [OpenTelemetry tracing specification](https://opentelemetry.io/docs/specs/otel/trace/api/) to ease compatibility with that ecosystem.
- But, not perfectly. I don't think wasi-observe should be 100% coupled to the OpenTelemetry ecosystem. For example here are just a few of the __many__ things we're not following.
  - "There MUST NOT be any API for creating a Span other than with a Tracer."
  - "[The new span API] MUST NOT accept a Span or SpanContext as parent, only a full Context."
- Liberally use `option<T>` rather than providing multiple versions of methods e.g. having `start-implicit: static func(name: string)` and `start-with-options: static func(name: string, options: start-options)` and ...

I'm happy to discuss the validity of these design principles and whether we want to take a different approach. They are not givens.

calebschoepp committed:
feat(wit): Propose WIT for tracing

Signed-off-by: Caleb Schoepp <caleb.schoepp@fermyon.com>

lann commented on wit/traces.wit:
```diff
@@ -0,0 +1,153 @@
+        /// Signals that the operation described by this span has now ended.
+        ///
+        /// If a timestamp is not provided then it is treated equivalent to passing the current time.
+        end: func(timestamp: option<datetime>);
```
Is dropping the resource equivalent to `end(none)`?

calebschoepp replied:
Yeah. That should probably be documented. I need to do some more thinking about the lifetimes of resources. B/c if we're forced to end a span when a resource is dropped then we might be breaking some of the OTel semantics.

jcchavezs replied:
I'd rather have a method call "abandon" instead.

calebschoepp replied:
Can you clarify what you'd want to have `abandon` be an alternative to?

lann commented on wit/traces.wit:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    /// A key-value pair describing an attribute.
+    record key-value {
```
```suggestion
    record attribute {
```

:shrug:

calebschoepp replied:
I also feel 🤷 on this. The Rust and Go OTel implementations call this `KeyValue`. But, the spec talks about it as an attribute. I think all things considered I would lean to sticking with `key-value`.  Especially given this line of the spec:

> An Attribute is a key-value pair

lann commented on wit/traces.wit:
```diff
@@ -0,0 +1,153 @@
+    /// Represents a unit of work or operation.
+    resource span {
+        /// Starts a new span with the given name, parent, and options.
+        start: static func(name: string, parent: span-parent, options: option<start-options>) -> span;
```
If `parent` is "almost always" `implicit` it could maybe move inside `options`? :shrug:

calebschoepp replied:
I like this

scb01 commented on wit/traces.wit:
```diff
@@ -0,0 +1,153 @@
+        trace-state: string,
+    }
+
+    /// The trace that this `span-context` belongs to.
+    type trace-id = tuple<u64, u64>;
+
+    /// The id of this `span-context`.
+    type span-id = u64;
```
I would prefer it to be a `string` as well.

jeffparsons replied:
> Use a fixed length array — Doesn't yet exist in WIT.

Might be worth a look at https://github.com/WebAssembly/component-model/pull/384 — it sounds like fixed-length arrays may exist soon?

scb01 commented on wit/traces.wit:
```diff
@@ -0,0 +1,153 @@
+        /// The `span`'s parent is the current active span. The current active span is the most recently created and non-closed span. If no spans have been started in the guest this may be a span in the host.
+        implicit,
+        /// The `span` is a root span and should have no parent.
+        is-root,
```
I'd like to discuss the implications of making the `span-parent` settable at `start`. Here is the way I understand it 

1. Implicit means that the span will appear in a logical, _expected_ place as a child of the previously executing span. 
2. `is-root` will detach this span from the previous one and will likely be the parent of a net new trace tree?
3. `span-context` will make this a child of a different span, possibly belong to a different trace tree. 

As of now, I find myself being the most uncomfortable about 3, as I'm unable to think of a usecase where this will apply.

calebschoepp replied:
You're understanding of the three options is correct. I largely designed it this way to follow the [spec](https://opentelemetry.io/docs/specs/otel/trace/api/#span-creation). However, I'm also starting to think that supporting options 2 and 3 is a mistake (at least for the first version of this WIT).

With option 2 I can sort of imagine a scenario where a user would want it. For example maybe they're running an app in Wasm that pulls work items off a queue. They might want the spans of pulling the work item to be part of the host trace — but then have the spans for actually doing the work be their own root span and thus separate trace.

With option 3 I can't imagine a way that it would be useful, but I can imagine lots and lots of ways that it could break things.

Overall I'm becoming skeptical that options 2 and 3 make sense in the context of a guest. We can always iterate so I'd prefer to start with something simple and would advocate we just remove the choice altogether and just always make it implicit.

scb01 replied:
Yeah, for option 2 an example was a scenario that I'd encountered previously, in which it was convenient to have the _initiator of a set of tasks_ and _the work done to fulfill_ per task be in separate traces. So, it has some utility. 

That said, I am onboard with keeping the wit simple by only supporting implicit for now and evolving from there.

calebschoepp commented on wit/traces.wit:
```diff
@@ -0,0 +1,153 @@
+        /// Whether this `span-context` was propagated from a remote parent.
+        is-remote: bool,
+        /// The `trace-state` for this `span-context`.
+        trace-state: string,
```
Good catch

calebschoepp commented on wit/traces.wit:
```diff
@@ -0,0 +1,153 @@
+        add-link: func(link: link);
+
+        /// Override the default `span` status, which is unset.
+        set-status: func(status: status);
```
I think the idea with `status` is that 90% of the time you should leave it unset. In the cases where you do set it, it can only be done later in the operation when you know its outcome. [Source](https://opentelemetry.io/docs/specs/otel/trace/api/#set-status).

calebschoepp commented on wit/traces.wit:
```diff
@@ -0,0 +1,153 @@
+        /// `link`'s for the new `span`.
+        links: option<list<link>>,
+        /// When the `span` should begin. If this is not provided it defaults to the current time.
+        timestamp: option<datetime>,
```
I personally view the dependency on `wasi::clocks` as not a very large downside and think that it is worth supporting setting these timestamps. But, I'm also curious on what the sentiment of the community is.

calebschoepp commented:
@chrisdickinson @brooksmtownsend wondering if either of you have had a chance to look at this.

calebschoepp committed:
Small changes

- Rename traces interface to tracer
- Change some types
- Move start to the interface rather than being a constructor
- Remove abillity to specify arbitrary span context as parent

Signed-off-by: Caleb Schoepp <caleb.schoepp@fermyon.com>

lann commented on wit/tracer.wit:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    /// Configuration for starting a `span`.
+    record start-options {
```
I'm wondering if this ought to be a resource to provide better forward compatibility for adding options.

calebschoepp replied:
Can you have fields on a resource or would all the data be stored in the host table?

lann replied:
It would all be stored by the resource owner, which would typically be the host

calebschoepp replied:
@brooksmtownsend what do you think about this?

lann commented on wit/tracer.wit:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    /// The `status` of a `span`.
+    variant status {
```
You could arguably represent this as an `option<result<_, string>>` :shrug:

fibonacci1729 replied:
+1 for `type status = option<result<_, string>>;`

lann replied:
I don't feel strongly about this one.

brooksmtownsend replied:
I think I agree with @calebschoepp here, I like the direct tie to the OTEL meaning

lann commented on wit/tracer.wit:
```diff
@@ -0,0 +1,151 @@
+        /// A boolean value.
+        %bool(bool),
+        /// A double precision floating point value.
+        %float64(float64),
+        /// A signed 64 bit integer value.
+        %s64(s64),
```
I'd maybe name these `float` and `int` :shrug:

fibonacci1729 replied:
+1

fibonacci1729 replied:
Also is it still `float64` or `f64` ?

calebschoepp replied:
Using `float64` is just an artifact of Spin being on an old wit-bindgen version.

calebschoepp commented on wit/tracer.wit:
```diff
@@ -0,0 +1,151 @@
+interface tracer {
```
Good call.

brooksmtownsend replied:
What do you think about just calling this `trace`? `wasi:observe/trace.start` feels nice. Don't want to bikeshed on the name too hard at this stage so feel free to just ignore

calebschoepp replied:
I originally had it named something very similar to that — `traces.wit`.

I changed it to `tracer.wit` on the premise that the interface itself is acting as the [`Tracer`](https://opentelemetry.io/docs/specs/otel/trace/api/#tracer). In the future if/when we support [WIT templating](https://github.com/WebAssembly/component-model/issues/172) then templating would act as the `[TracerProvider`](https://opentelemetry.io/docs/specs/otel/trace/api/#tracerprovider).

I imagine that we would also have a similar pattern for `meter.wit` and `logger.wit`.

brooksmtownsend commented on wit/tracer.wit:
```diff
@@ -0,0 +1,151 @@
+
+    /// A key-value pair describing an attribute.
+    record key-value {
+        /// The attribute name.
+        key: key,
```
nit: I generally prefer not using simple type aliases in WIT since it's just another mental lookup, thoughts on just making this a `string`?

brooksmtownsend replied:
I have the same thoughts later about `trace-id` / `span-id` etc etc but just to keep the comment spread simple I'll raise that here

calebschoepp replied:
`key` being type aliased was very much a result of me closely mirroring the OTel spec. Thinking about it some more I agree that we should just remove this alias and have `key-value` defined as:
```wit
/// A key-value pair describing an attribute.
record key-value {
    /// The attribute name.
    key: string,
    /// The attribute value.
    value: value,
}
```

As for `trace-id` / `span-id`, good catch. I also think that those should no longer be type aliased. Earlier they were some weird data structure like `tuple<u64, u64>` which made an alias more relevant.

```wit
/// Identifying trace information about a span that can be serialized and propagated.
record span-context {
    /// The `trace-id` for this `span-context`.
    trace-id: string,
    /// The `span-id` for this `span-context`.
    span-id: string,
...
}
```

brooksmtownsend replied:
That sounds good! Looks like I was missing a bit of history but glad it's landing in a simpler place 😄

brooksmtownsend commented on wit/tracer.wit:
```diff
@@ -0,0 +1,151 @@
+        /// If a key already exists for an attribute of the Span it will be overwritten with the corresponding new value.
+        set-attributes: func(attributes: list<key-value>);
+
+        /// Adds an event with the provided name at the curent timestamp.
```
nit
```suggestion
        /// Adds an event with the provided name at the current timestamp.
```

calebschoepp replied:
I'm a little surprised this made it past my VS Code spelling extension

brooksmtownsend commented on wit/tracer.wit:
```diff
@@ -0,0 +1,151 @@
+        attributes: option<list<key-value>>,
+        /// `link`'s for the new `span`.
+        links: option<list<link>>,
+        /// When the `span` should begin. If this is not provided it defaults to the current time.
+        timestamp: option<datetime>,
```
Is this a common use case, to delay beginning of a span? I'm wondering if this would be hard to do / introduce weird cases where components aren't able to retrieve the current time, so then this timestamp doesn't work as expected.

calebschoepp replied:
> Is this a common use case
🤷 I have no clue. It'd be really good to here from end users how important this functionality is.

I definitely share some of the same wariness about the edge cases this could introduce. If we made `start-options` a resource like @lann suggested then it would be easy to punt on this functionality for v0 and then add it in the future if it is important to people.

> where components aren't able to retrieve the current time
I'm not really sure how this would ever happen. Mind expanding?

brooksmtownsend replied:
> where components aren't able to retrieve the current time

Oh yeah, I only assumed that if a component was going to start a span at a later or earlier time, it was going to be relative to whatever it could fetch via `wasi:clocks`. If a runtime changed the implementation of `wasi:clocks` to either disallow fetching or just giving back a time that's very low resolution, it would just create inaccurate traces.

calebschoepp replied:
I see. Yep that would definitely be a failure case. But, I'm not really sure that is a WASI Observe problem. In the case where someone has virtualized `wasi:clocks` to be low resolution aren't all bets off anywhere you want to use those time values?

Very open to changing my mind here so I'm curious what you think.

brooksmtownsend replied:
>  In the case where someone has virtualized wasi:clocks to be low resolution aren't all bets off anywhere you want to use those time values?

Yep, pretty much I think. That's mostly why I was wondering if it was a common enough use case but I think I can see the utility. If the options gets turned into a Resource this may be a good thing to leave off, but if not I don't think it's a big deal 🙂

calebschoepp replied:
Cool. I'm onboard with leaving it off it we make options a resource (which I think we should do)

brooksmtownsend commented on wit/tracer.wit:
```diff
@@ -0,0 +1,151 @@
+        new-root: bool,
+        /// `span-kind` for the new `span`.
+        span-kind: option<span-kind>,
+        /// Attributes for the new `span`.
+        attributes: option<list<key-value>>,
+        /// `link`'s for the new `span`.
+        links: option<list<link>>,
```
What do you think about this just being a `list<..>` instead of an `option<list<..>>`? Is providing an empty list of attributes semantically different from no list?

calebschoepp replied:
There is no semantic difference. The reason I did this is for code gen to be nicer in languages that allow you to not define everything in a struct like JavaScript. In JS this would let you do:

```js
tracer.start("foo", startOptions{ newRoot: true })
```

Instead of:

```js
tracer.start("foo", startOptions{ newRoot: true, links: [], ... })
```

brooksmtownsend replied:
That's good enough for me given the nice handling of optionals here

brooksmtownsend reviewed (commented):
Hey @calebschoepp, fantastic job here! I really like the direct ties to the traces page from the otel docs. Feels like a great approach.

I had mostly nits and a couple of comments that I'm interested in and happy to talk sync in the next meeting too.

(p.s. thank you for the patience on the review here, I lost this in my inbox)
